crates/sui-framework/packages/dwallet-system/sources/ethereum_state.move

```diff
@@ -64,28 +75,88 @@ module dwallet_system::ethereum_state {
         updates_bytes: vector<u8>,
         finality_update_bytes: vector<u8>,
         optimistic_update_bytes: vector<u8>,
-        state_bytes: vector<u8>,
         latest_ethereum_state: &mut LatestEthereumState,
+        eth_state:  &EthereumState,
         ctx: &mut TxContext,
     ) {
-        let (data, time_slot) = verify_eth_state(
+        // Verify that the state is the latest state
+        let eth_state_id = object::id(eth_state);
+        assert!(eth_state_id == latest_ethereum_state.eth_state_id, EStateObjectMismatch);
+
+        // Verify that the state is based on the latest state
+        let latest_ethereum_state_id = get_ethereum_state_latest_state_id(eth_state);
+        assert!(latest_ethereum_state_id == object::id(latest_ethereum_state), EStateObjectMismatch);
+
+        let eth_state_bytes = get_ethereum_state_data(eth_state);
+        let (data, time_slot, state_root, _network) = verify_eth_state(
             updates_bytes,
             finality_update_bytes,
             optimistic_update_bytes,
-            state_bytes
+            eth_state_bytes
         );
 
-        let new_state = EthereumState {
-            id: object::new(ctx),
-            data,
-            time_slot,
-        };
+        if (time_slot > latest_ethereum_state.last_slot) {
```

Comment:
let's move this `if` to the beginning of the function, and actually make it an `assert`. No other purpose to call this function